### PR TITLE
fix: implement soft delete pattern for StationConnection to eliminate 503 window during network graph rebuilds

### LINKED ISSUE #230: 503 error viewing routes when backend + celery started
does not happen when only fastapi is started. This is because on startup the entire tube network is refreshed (implemented in #218) but it does this by blocking the database or deleting the database -> confirmed by commenting out the task in app.celery.app.trigger_startup_tasks()

The refresh of the tube network should not block users accessing the app. It should happen transparently - perhaps using the `deleted_at` and `updated_at` attribute in the affected tables to allow the existing dataset to be used while a new dataset is inserted and then an atomic action will activate the changes. This would also allow us to track changes in what is supposed to be a highly static network. 

backend:
```python
IsTheTubeRunning/backend on  main [$] via 🐳 colima is 📦 v0.1.0 via 🐍 v3.12.8 on ☁️  robert.aleck@cynexia.com took 5m36s 
❯  cd /Users/rob/Downloads/git/IsTheTubeRunning/backend ; /usr/bin/env /Users/rob/Downloads/git/IsTheTubeRunning/backend/.venv/bin/python /Users/rob/.vscode/extensions/ms-python.debugpy-2025.16.0-darwin-arm64/bundled/libs/debugpy/adapter/../../debugpy/laun
cher 52937 -- -m uvicorn app.main:app --reload --host 0.0.0.0 --port 8000 
INFO:     Will watch for changes in these directories: ['/Users/rob/Downloads/git/IsTheTubeRunning/backend']
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)
INFO:     Started reloader process [41847] using WatchFiles
2025-11-21T09:27:59.478929Z [info     ] celery_otel_instrumentation_enabled [app.celery.app]
2025-11-21T09:27:59.635006Z [warning  ] sms_file_logging_disabled      [app.services.sms_service] reason='SMS_LOG_DIR not configured'
2025-11-21T09:28:00.198542Z [info     ] otel_fastapi_instrumented      [app.main] excluded_urls=['/health', '/ready', '/metrics']
INFO:     Started server process [42058]
INFO:     Waiting for application startup.
2025-11-21T09:28:00.234503Z [info     ] otel_tracer_provider_created   [app.core.telemetry] endpoint=https://otlp-gateway-prod-gb-south-1.grafana.net/otlp/v1/traces environment=dev service_name=isthetuberunning-api
2025-11-21T09:28:00.234663Z [info     ] otel_tracer_provider_initialized [app.main]
2025-11-21T09:28:00.234775Z [info     ] startup_initializing           [app.main] message='validating database'
2025-11-21T09:28:00.351258Z [info     ] database_connection_successful [app.main]
2025-11-21T09:28:00.351665Z [info     ] Context impl PostgresqlImpl.   [alembic.runtime.migration]
2025-11-21T09:28:00.351736Z [info     ] Will assume transactional DDL. [alembic.runtime.migration]
2025-11-21T09:28:00.382069Z [info     ] database_migration_valid       [app.main] revision=7df10ab411af
2025-11-21T09:28:00.425373Z [info     ] severity_codes_cache_warmed    [app.services.tfl_service] count=233 ttl=86400
2025-11-21T09:28:00.426616Z [info     ] disruption_categories_cache_warmed [app.services.tfl_service] count=7 ttl=86400
2025-11-21T09:28:00.427565Z [info     ] stop_types_cache_warmed        [app.services.tfl_service] count=3 ttl=86400
2025-11-21T09:28:00.428211Z [info     ] redis_cache_warmup_complete    [app.main] disruption_categories=7 severity_codes=233 stop_types=3
2025-11-21T09:28:00.429148Z [info     ] startup_complete               [app.main]
INFO:     Application startup complete.
...
2025-11-21T09:28:08.145491Z [info     ] fetching_network_graph         [app.services.tfl_service] span_id=f9de91d3abe21ef8 trace_id=9117046473e4b6202ff7dd04671b7b8a
2025-11-21T09:28:08.148343Z [info     ] http_request                   [app.middleware.access_logging] client_ip=127.0.0.1 duration_ms=160.39 method=GET path=/api/v1/tfl/lines span_id=9106f997714c67a9 status_code=200 trace_id=976905a191ea9f06c023c564bee982d2
2025-11-21T09:28:08.153034Z [info     ] http_request                   [app.middleware.access_logging] client_ip=127.0.0.1 duration_ms=17.02 method=GET path=/api/v1/tfl/network-graph span_id=f9de91d3abe21ef8 status_code=503 trace_id=9117046473e4b6202ff7dd04671b7b8a
...
```

### PR DESCRIPTION
**Pull request by mnbf9rca**
resolves #230

## Summary by Sourcery

Implement a soft delete strategy for station network connections to avoid downtime during network graph rebuilds and document the new approach.

Bug Fixes:
- Eliminate the 503 error window during network graph rebuilds by keeping existing station connections visible until new ones are committed.

Enhancements:
- Adopt a partial unique index on active station connections to support atomic soft deletes and rebuilds.
- Ensure all station connection queries and helpers only consider non-deleted records for network graph operations.
- Re-enable the startup Celery task that rebuilds the network graph now that rebuilds are zero-downtime.

Build:
- Add an Alembic migration to replace the station connection unique constraint with a partial unique index scoped to non-deleted records.

Documentation:
- Document the soft delete pattern with a partial unique index for network graph rebuilds in the external APIs ADR, including rationale and consequences.

Tests:
- Add dedicated tests to verify isolation behavior during network graph rebuilds and correct handling of soft-deleted connections.
- Update existing unit and integration tests to assert against active (non-deleted) station connections only.

### CODE REVIEW

**Comment by mnbf9rca** [comment]
Thanks for the thoughtful review! I've addressed the accepted suggestions and here's my rationale for the other comments:

**Encapsulating `deleted_at.is_(None)` filter:**
Following the YAGNI principle, this filter is currently used in only a few locations within the TfL service, all well-documented with the new ADR. There's no evidence of accidental omissions. If we add more soft-deleted models in the future, we'll consider a base class or mixin pattern, but for now the current approach maintains simplicity appropriate for a hobby project.

**Using soft-delete mixin helper for tests:**
No soft-delete mixin currently exists in the codebase. The soft delete pattern was just implemented specifically for `StationConnection`. Tests directly setting the `deleted_at` field is the most straightforward approach and correctly tracks deletion behavior.

**Exposing `_connection_exists` as public method:**
`_connection_exists` is an internal helper for route validation with no user-facing use case. Testing private methods is appropriate when they contain complex logic that needs verification. There's no need to expose this as a public API endpoint.

**Adding integration test for zero-downtime:**
The existing unit tests adequately cover the soft delete behavior. Testing concurrent transactions would be complex and potentially flaky. PostgreSQL's transaction semantics provide the isolation guarantee documented in the ADR. For a hobby project, the current test coverage is sufficient.

**Changes made:**
- ✅ Changed `datetime.now(UTC)` to `func.now()` for database-side timestamp consistency
- ✅ Fixed typo in ADR ("becomes issue" → "becomes an issue")

**Comment by mnbf9rca** [comment]
## PR Review Comment Resolution Plan

### Summary
Addressed PR review comments from Sourcery AI bot. Made 2 code changes based on accepted suggestions and provided rationale for 4 rejected suggestions.

### Comment Decisions

#### ✅ ACCEPTED (2 comments)
1. **Use `func.now()` for database-side timestamp** (backend/app/services/tfl_service.py:2584)
   - Changed `datetime.now(UTC)` → `func.now()`
   - Rationale: Best practice for consistency with DB timestamps
   - Status: ✅ Complete

2. **Fix typo in ADR** (docs/adr/07-external-apis.md:410)
   - Changed "becomes issue" → "becomes an issue"
   - Rationale: Grammar correction
   - Status: ✅ Complete

#### ❌ REJECTED (4 comments - with rationale)
1. **Encapsulate `deleted_at.is_(None)` filter**
   - Rationale: YAGNI - only used in few places, well-documented, no evidence of omissions
   
2. **Use soft-delete mixin helper for tests**
   - Rationale: No mixin exists; direct field manipulation is straightforward
   
3. **Expose `_connection_exists` as public method**
   - Rationale: Internal helper with no user-facing need; testing private methods is appropriate
   
4. **Add integration test for zero-downtime**
   - Rationale: Existing tests adequate; concurrent transaction testing would be complex/flaky

### Testing
- ✅ Network graph rebuild isolation tests: 4/4 passed
- ✅ Build station graph tests: 5/5 passed  
- ✅ All pre-commit hooks passed

### Files Modified
- `backend/app/services/tfl_service.py` (line 2584)
- `docs/adr/07-external-apis.md` (line 410)

### Next Steps
Ready for user to push changes. Once pushed, will reply to accepted comments with commit SHA.

File: backend/app/services/tfl_service.py [hunk 1]
```diff
@@ -2574,12 +2575,18 @@ async def build_station_graph(self) -> dict[str, int]:
...
-            # Without this, SQLAlchemy may reorder operations causing duplicate key errors.
-            await self.db.flush()
-            logger.info("existing_connections_cleared")
+            # Soft delete existing connections (within transaction - will rollback if building fails)
+            # Use soft delete instead of hard delete to eliminate 503 window during rebuild.
+            # Old connections remain visible until new connections are committed (issue #230).
+            await self.db.execute(
+                update(StationConnection)
+                .where(StationConnection.deleted_at.is_(None))
+                .values(deleted_at=datetime.now(UTC))
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/app/services/tfl_service.py]: ✅ Fixed in commit 3965bc3e8f9278f55f0f428260e6d2f22faff177

Changed `datetime.now(UTC)` to `func.now()` for database-side timestamp consistency. Tests passed and pre-commit hooks verified.

File: docs/adr/07-external-apis.md [hunk 2]
```diff
@@ -348,3 +348,67 @@ Database-only for public endpoints: validate line exists before database query,
...
+- Atomic transition from old to new graph data
+- No flush() needed - true ACID transaction
+- Startup rebuild can be re-enabled safely
+
+**More Difficult:**
+- Table accumulates soft-deleted records over time (one full graph per rebuild)
+- All `StationConnection` queries must filter `deleted_at.is_(None)` for correctness
+- Slightly larger table size (mitigated: daily rebuilds = ~365 old graphs per year = minimal)
+- Requires partial index migration (PostgreSQL-specific feature)
+- Future cleanup task may be needed if growth becomes issue (deferred as YAGNI)
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply docs/adr/07-external-apis.md]: ✅ Fixed in commit 3965bc3e8f9278f55f0f428260e6d2f22faff177

Updated ADR to fix grammar: "becomes issue" → "becomes an issue"

**Comment by mnbf9rca** [comment]
## ✅ All Review Comments Addressed

All PR review comments have been addressed and resolved.

### Changes Made (Commit: 3965bc3)
1. ✅ Changed `datetime.now(UTC)` to `func.now()` in backend/app/services/tfl_service.py:2584
2. ✅ Fixed typo in docs/adr/07-external-apis.md:410 ("becomes issue" → "becomes an issue")

### Testing
- ✅ Network graph rebuild isolation tests: 4/4 passed
- ✅ Build station graph tests: 5/5 passed
- ✅ All pre-commit hooks passed
- ✅ No regressions introduced

### Comment Resolution
- ✅ 2 accepted comments resolved with commit SHA
- ✅ 4 rejected comments with documented rationale
- ✅ Both review threads marked as resolved

**PR is ready for merge.**
